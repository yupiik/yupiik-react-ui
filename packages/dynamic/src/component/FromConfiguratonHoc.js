import React from 'react';
import { useEffect, useMemo, useReducer } from "react";
import { Dynamic } from "./Dynamic";
import jsonLogic from './json-logic.operators';

const rewriteProp = (source, args) => {
    if (Array.isArray(source)) {
        return source.map(it => typeof it === 'object' ? rewriteProp(it, args) : it);
    }
    if (source.jsonLogic) {
        return jsonLogic.apply(source.jsonLogic, args);
    }
    return source;
};

/**
 * A HoC wrapping {@link Dynamic} component with some more connection options.
 * Its primary intent is to provide a plain configuration (JSON friendly) to render support - whereas dynamic still requires javascript for callacks for example.
 * For callbacks, it uses JSON-Logic evaluation to solve the JSON -> js mapping.
 * 
 * @param {object} registry component registry (key matching the component name)
 * @param {string} name the component name in the registry
 * @param {array} useEffectCallback the {@link useEffect} callbacks properties definition in JSON-Logic format, it can be an array or an object (behaves as an array of 1 element). Each element is a JSON-Logic expression.
 * @param {array} useReducerCallback the {@link useReducer} callbacks. {@constant initialState} is the state initial value and {@constant reducerJsonLogic} the JSON-Logic implementation which takes the current state and action as parameters. JSONLogic take the state and dispatch functions as parameters.
 * @param {object} callbacks the callbacks properties definition in JSON-Logic format, key is the callback name and value its JSON-Logic chain. The JSON-Logic can use variables `state`, `dispatch` and `event` (the callback event).
 * @param {object} configuration properties to passthrough the underlying component.
 * @param {object} parentState if no {@constant initialState} is passed it will be used to overwrite it, enables to forward through children the state when children are compatible.
 * @param {object} parentDispatch parent dispatch callback (to be able to modify parent state).
 */
export const FromConfiguratonHoc = ({
    registry = undefined,
    parentState = undefined,
    parentDispatch = undefined,
    options: {
        name,
        keepExtendedProps = false, // if some wrapper injects properties, forward them to the wrappped components (routers often do that, conditional does too)
        options: {
            useReducerCallback: {
                initialState = undefined,
                reducerJsonLogic = false,
            } = {},
            useEffectCallback = [],
            callbacks = {},
            configuration = {},
        } = {},
    },
    ...rest
}) => {
    const [state, dispatch] = useReducer(
        (currentState, action) => {
            if (action.type === '$state') {
                return action.value;
            }
            if (reducerJsonLogic) {
                return jsonLogic.apply(reducerJsonLogic, { parentState, parentDispatch, state: currentState, action });
            }
            if (parentDispatch) { // bubble up if there is no own definition
                parentDispatch(action);
            }
            // here there is no reducerJsonLogic so we just back with parent state
            return parentState;
        },
        initialState,
        localState => localState === undefined ? parentState : localState);

    useEffect(() => {
        if (!reducerJsonLogic) {
            dispatch({ type: '$state', value: parentState });
        }

        if (!useEffectCallback) {
            return;
        }

        let result;
        if (Array.isArray(useEffectCallback)) {
            result = useEffectCallback
                .map(it => jsonLogic.apply(it, { parentState, parentDispatch, state, dispatch }))
                .filter(it => it)
                .reduce((a, i) => {
                    if (typeof i === 'function') {
                        return () => {
                            a();
                            i();
                        };
                    }
                }, () => { });
        } else {
            result = jsonLogic.apply(useEffectCallback, { parentState, parentDispatch, state, dispatch });
        }
        if (typeof result === 'function') {
            return result;
        }
    }, [useEffectCallback, initialState, parentState]);

    const computedOptions = useMemo(() => {
        const opts = Object
            .keys(configuration)
            .reduce(
                (aggregator, name) => {
                    const value = configuration[name];
                    aggregator[name] = typeof value === 'object' ?
                        rewriteProp(value, {
                            parentState,
                            parentDispatch,
                            state,
                            dispatch,
                            props: rest,
                        }) :
                        value;
                    return aggregator;
                },
                !keepExtendedProps ? {} : { ...(rest || {}) });

        const callbackOpts = Object
            .keys(callbacks)
            .reduce((aggregator, callback) => {
                const spec = callbacks[callback];
                aggregator[callback] = e => jsonLogic.apply(spec, {
                    parentState,
                    parentDispatch,
                    state,
                    dispatch,
                    event: e,
                    props: rest,
                });
                return aggregator;
            }, {});

        return { name, options: { ...opts, ...callbackOpts, state, dispatch } };
    }, [callbacks, configuration, state, parentState, rest]);

    return (
        <Dynamic
            registry={registry}
            options={computedOptions}
        />
    );
};
