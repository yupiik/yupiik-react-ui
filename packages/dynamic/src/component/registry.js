import { Component, cloneElement } from 'react';
import { FromConfiguratonHoc } from './FromConfiguratonHoc';

export const simpleComponent = nameOrComponent => ({
    registry,
    children,
    state,
    dispatch,
    ...props
}) => {
    const Comp = nameOrComponent;
    const subComponents = children !== undefined && (Array.isArray(children) ? children : [children])
        .filter(it => it !== undefined)
        .map(it => {
            // all the trick is there, wrap the children with FromConfiguratonHoc to enable the configuration to propagate properly 
            if (it && typeof it === 'object' && it.options) {
                if (it.name) { // not wrapped, we tolerate it since it is simpler to write
                    let otherProps = it.options.wrapperProps || {};
                    if (it.keepExtendedProps && props) {
                        otherProps = { ...otherProps, ...props };
                    }
                    return (
                        <FromConfiguratonHoc
                            {...otherProps}
                            registry={registry}
                            options={it}
                            parentState={state}
                            parentDispatch={dispatch}
                        />
                    );
                }
                return (
                    <FromConfiguratonHoc registry={registry} {...it} parentState={state} parentDispatch={dispatch} />
                );
            }
            return it;
        });
    if (subComponents && subComponents.length > 0) {
        return (
            <Comp {...props}>
                {subComponents}
            </Comp>
        );
    }
    return (<Comp {...props} />)
};

const isComponent = it =>
    it.charAt(0) === it.charAt(0).toUpperCase() &&
    !it.includes('$') &&
    !it.includes('_') &&
    !it.endsWith('Props');

/**
 * Create a registry from immediately nested entries.
 * @param {object} an import of components.
 * @returns a registry wrapping components.
 */
export const simpleRegistry = object => Object
    .keys(object)
    .reduce((a, i) => {
        a[i] = simpleComponent(object[i]);
        return a;
    }, {});

/**
 * Create a registry from immediately nested entries and their children.
 * @param {object} an import of components.
 * @returns a registry wrapping components.
 */
export const nestedRegistry = object => Object
    .keys(object)
    .filter(it => isComponent(it))
    .reduce((a, i) => {
        const wrapper = object[i];

        a[i] = simpleComponent(wrapper);

        // subcomponents
        Object
            .keys(wrapper)
            .filter(it => isComponent(it))
            .forEach(it => {
                a[`${i}.${it}`] = simpleComponent(wrapper[it]);
            });

        return a;
    }, {});

// will generally be overriden by apps but enables to get a default
class SimpleErrorBoundary extends Component {
    state = { error: null }

    static getDerivedStateFromError(error) {
        return { error: error.message }
    }

    componentDidCatch(error) {
        console.error(error)
        this.setState({ error: error.message })
    }

    render() {
        if (this.state.error) {
            return <p>Oh no! An error occured: {this.state.error}</p>
        }
        return this.props.children;
    }
}

const Conditional = ({ children, condition, ...rest }) => {
    return condition && children.map(it => {
        return cloneElement(it, rest || {});
    });
};

export const HtmlRegistry = {
    a: simpleComponent('a'),
    abbr: simpleComponent('abbr'),
    address: simpleComponent('address'),
    area: simpleComponent('area'),
    article: simpleComponent('article'),
    aside: simpleComponent('aside'),
    audio: simpleComponent('audio'),
    b: simpleComponent('b'),
    base: simpleComponent('base'),
    bdi: simpleComponent('bdi'),
    bdo: simpleComponent('bdo'),
    blockquote: simpleComponent('blockquote'),
    body: simpleComponent('body'),
    br: simpleComponent('br'),
    button: simpleComponent('button'),
    canvas: simpleComponent('canvas'),
    caption: simpleComponent('caption'),
    cite: simpleComponent('cite'),
    code: simpleComponent('code'),
    col: simpleComponent('col'),
    colgroup: simpleComponent('colgroup'),
    data: simpleComponent('data'),
    datalist: simpleComponent('datalist'),
    dd: simpleComponent('dd'),
    del: simpleComponent('del'),
    details: simpleComponent('details'),
    dfn: simpleComponent('dfn'),
    dialog: simpleComponent('dialog'),
    div: simpleComponent('div'),
    dl: simpleComponent('dl'),
    dt: simpleComponent('dt'),
    em: simpleComponent('em'),
    embed: simpleComponent('embed'),
    fieldset: simpleComponent('fieldset'),
    figcaption: simpleComponent('figcaption'),
    figure: simpleComponent('figure'),
    footer: simpleComponent('footer'),
    form: simpleComponent('form'),
    h1: simpleComponent('h1'),
    h2: simpleComponent('h2'),
    h3: simpleComponent('h3'),
    h4: simpleComponent('h4'),
    h5: simpleComponent('h5'),
    h6: simpleComponent('h6'),
    head: simpleComponent('head'),
    header: simpleComponent('header'),
    hr: simpleComponent('hr'),
    html: simpleComponent('html'),
    i: simpleComponent('i'),
    iframe: simpleComponent('iframe'),
    img: simpleComponent('img'),
    input: simpleComponent('input'),
    ins: simpleComponent('ins'),
    kbd: simpleComponent('kbd'),
    label: simpleComponent('label'),
    legend: simpleComponent('legend'),
    li: simpleComponent('li'),
    link: simpleComponent('link'),
    main: simpleComponent('main'),
    map: simpleComponent('map'),
    mark: simpleComponent('mark'),
    meta: simpleComponent('meta'),
    meter: simpleComponent('meter'),
    nav: simpleComponent('nav'),
    object: simpleComponent('object'),
    ol: simpleComponent('ol'),
    optgroup: simpleComponent('optgroup'),
    option: simpleComponent('option'),
    p: simpleComponent('p'),
    param: simpleComponent('param'),
    picture: simpleComponent('picture'),
    pre: simpleComponent('pre'),
    progress: simpleComponent('progress'),
    q: simpleComponent('q'),
    rp: simpleComponent('rp'),
    rt: simpleComponent('rt'),
    ruby: simpleComponent('ruby'),
    s: simpleComponent('s'),
    samp: simpleComponent('samp'),
    script: simpleComponent('script'),
    section: simpleComponent('section'),
    select: simpleComponent('select'),
    small: simpleComponent('small'),
    source: simpleComponent('source'),
    span: simpleComponent('span'),
    strong: simpleComponent('strong'),
    style: simpleComponent('style'),
    sub: simpleComponent('sub'),
    summary: simpleComponent('summary'),
    sup: simpleComponent('sup'),
    svg: simpleComponent('svg'),
    table: simpleComponent('table'),
    tbody: simpleComponent('tbody'),
    td: simpleComponent('td'),
    template: simpleComponent('template'),
    textarea: simpleComponent('textarea'),
    tfoot: simpleComponent('tfoot'),
    th: simpleComponent('th'),
    thead: simpleComponent('thead'),
    time: simpleComponent('time'),
    title: simpleComponent('title'),
    tr: simpleComponent('tr'),
    track: simpleComponent('track'),
    u: simpleComponent('u'),
    ul: simpleComponent('ul'),
    var: simpleComponent('var'),
    video: simpleComponent('video'),
    wbr: simpleComponent('wbr'),

    // not strictly html but from "core"
    ErrorBoundary: simpleComponent(SimpleErrorBoundary),
    FromConfiguratonHoc,
    Conditional: simpleComponent(Conditional),
};
