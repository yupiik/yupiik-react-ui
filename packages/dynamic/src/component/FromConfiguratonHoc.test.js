import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import { expect } from 'expect';
import { HtmlRegistry } from './registry';
import { FromConfiguratonHoc } from './FromConfiguratonHoc';

const TestFCHoc = (props) => (
    <FromConfiguratonHoc
        registry={HtmlRegistry}
        options={props}
    />
);

test('FromConfiguratonHoc should render static components', () => {
    const { container: { children } } = render(
        <TestFCHoc
            name="div"
            options={{
                callbacks: {},
                configuration: {
                    className: 'test',
                    children: 'Tests',
                },
            }}
        />
    );
    expect(children).toMatchSnapshot();
});

// inspired from
// https://preactjs.com/guide/v10/hooks/#usereducer
// and
// https://preactjs.com/guide/v10/preact-testing-library/#usage
test('FromConfiguratonHoc should render components with state', async () => {
    const { container: { children } } = render(
        <TestFCHoc
            name="div"
            options={{
                useReducerCallback: {
                    initialState: 0,
                    /* equivalent to
                    const reducer = (state, action) => {
                    switch (action) {
                        case 'increment': return state + 1;
                        case 'decrement': return state - 1;
                        case 'reset': return 0;
                        default: throw new Error('Unexpected action');
                    }
                    };
                    */
                    reducerJsonLogic: {
                        'if': [
                            { '==': [{ 'var': 'action' }, 'increment'] },
                            { '+': [{ 'var': 'state' }, 1] },
                            {
                                'if': [
                                    { '==': [{ 'var': 'action' }, 'decrement'] },
                                    { '-': [{ 'var': 'state' }, 1] },
                                    {
                                        'if': [
                                            { '==': [{ 'var': 'action' }, 'reset'] },
                                            0,
                                            // not a real operator but will fail and it is what we want
                                            { 'invalid_action': [] },
                                        ],
                                    },
                                ],
                            },
                        ],
                    },
                },
                configuration: {
                    className: 'test',
                    children: [
                        { // <span>Current value: {count}</span>
                            name: 'span',
                            options: {
                                configuration: {
                                    'data-testid': 'current',
                                    children: [
                                        {
                                            jsonLogic: {
                                                cat: [
                                                    'Current value: ',
                                                    { var: 'state' },
                                                ]
                                            },
                                        },
                                    ],
                                },
                            },
                        },
                        { // <button onClick={() => dispatch('increment')}>+1</button>
                            name: 'button',
                            options: {
                                callbacks: {
                                    onClick: {
                                        'fn': [
                                            { var: 'dispatch' },
                                            'increment'
                                        ],
                                    },
                                },
                                configuration: {
                                    children: ['+1'],
                                },
                            },
                        },
                        { // <button onClick={() => dispatch('decrement')}>-1</button>
                            name: 'button',
                            options: {
                                callbacks: {
                                    onClick: {
                                        'fn': [
                                            { var: 'dispatch' },
                                            'decrement',
                                        ],
                                    },
                                },
                                configuration: {
                                    children: ['-1'],
                                },
                            },
                        },
                        { // <button onClick={() => dispatch('reset')}>reset</button>
                            name: 'button',
                            options: {
                                callbacks: {
                                    onClick: [
                                        {
                                            'fn': [
                                                { var: 'dispatch' },
                                                'reset',
                                            ],
                                        }
                                    ],
                                },
                                configuration: {
                                    children: ['Reset'],
                                },
                            },
                        }
                    ],
                },
            }}
        />
    );

    expect(children).toMatchSnapshot();

    fireEvent.click(screen.getByText('+1'));
    await waitFor(() => expect(screen.getByTestId('current').innerHTML).toBe('Current value: 1'));
});
