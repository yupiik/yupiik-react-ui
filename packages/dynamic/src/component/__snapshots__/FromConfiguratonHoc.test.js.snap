// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FromConfiguratonHoc should render components with state 1`] = `
HTMLCollection [
  <div
    class="test"
  >
    <span
      data-testid="current"
    >
      Current value: 0
    </span>
    <button>
      +1
    </button>
    <button>
      -1
    </button>
    <button>
      Reset
    </button>
  </div>,
]
`;

exports[`FromConfiguratonHoc should render static components 1`] = `
HTMLCollection [
  <div
    class="test"
  >
    Tests
  </div>,
]
`;
